### PR TITLE
Link analysis

### PR DESCRIPTION
**Pull request by st-alice**
# Summary

PR related to query-dependent link analysis to determine Start a Business pages. This requires Gdrive authentication (using service account credentials), and uses the output from the Bigquery script. 

# Checklists

This pull/merge request meets the following requirements:

- [X] code runs
- [X] [developments are ethical][data-ethics-framework] and secure
- [X] you have made proportionate checks that the code works correctly
~~- [ ] test suite passes~~
~~- [ ] developments adhere to AQA plan (see `docs/aqa/aqa_plan.md`)~~
~~- [ ] data log updated (see `docs/aqa/data_log.md`), if necessary~~
~~- [ ] assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if
  necessary~~
~~- [ ] [minimum usable documentation][agilemodeling] written in the `docs` folder~~

documentation will be updated in a separate PR. 

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: notebooks/query-dependent link analysis.ipynb [hunk 1]
```diff
@@ -0,0 +1,1064 @@
...
+   "id": "90df4bb9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Which nodes are in the list but not in the graph? \n",
+    "## Answer: two pages in the list no longer exist, so are not in the graph\n",
+    "\n",
+    "# List of node property 'name'\n",
+    "nodes_graph = []\n",
+    "for k,v in g.nodes(data=True):\n",
```
Comment (JakeRutherford) [inline notebooks/query-dependent link analysis.ipynb]: Try list comprehension:

nodes_graph = [v['properties']['name'] for _,v in g.nodes(data=True)]

File: notebooks/query-dependent link analysis.ipynb [hunk 2]
```diff
@@ -0,0 +1,1064 @@
...
+    "g_di = nx.DiGraph(g)\n",
+    "g_di_between = nx.betweenness_centrality(g_di)\n",
+    "g_between = dict(sorted(g_di_between.items(), key=operator.itemgetter(1),reverse=True))\n",
+    "\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name']) # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
```
Comment (JakeRutherford) [inline notebooks/query-dependent link analysis.ipynb]: Can this all be re written as:

node_names_dict = {k:v['properties']['name'] for k,v in g.nodes(data=True)}

??

File: notebooks/query-dependent link analysis.ipynb [hunk 3]
```diff
@@ -0,0 +1,1064 @@
...
+    "g_degree = dict(sorted(g_degree.items(), key=operator.itemgetter(1),reverse=True))\n",
+    "\n",
+    "[node[1]['properties']['name'] for node in G.nodes(data=True)]\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name'])  # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
+    "g_degree = {v: node_names_dict.get(k, v) for k, v in g_degree.items()}  # replace cid with node name \n",
```
Comment (JakeRutherford) [inline notebooks/query-dependent link analysis.ipynb:333]: As above, and again for similar bits of code like this below

File: notebooks/query-dependent link analysis.ipynb [hunk 4]
```diff
@@ -0,0 +1,1064 @@
...
+   "id": "50bbf9da",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Which nodes have user_movement in or out? 186\n",
+    "user_movement_nodes = []\n",
+    "\n",
+    "for node1, node2, data in g.edges(data=True):\n",
+    "    if data['type']=='USER_MOVEMENT':\n",
+    "            user_movement_nodes.append(node1)\n",
```
Comment (JakeRutherford) [inline notebooks/query-dependent link analysis.ipynb]: Can this all be rewritten as:

user_movement_nodes = [node1 for node1,node2,data in g.edges(data=True) if data['type'] == 'USER_MOVEMENT']

**Review by JakeRutherford — COMMENTED** [review]
Thanks Alice. Just needs some cleaner code in areas, using list comprehensions and dict comprehensions.

File: notebooks/query-dependent link analysis.ipynb (hunk 1 above)
Comment (st-alice) [reply notebooks/query-dependent link analysis.ipynb]: Thank you! Refactored in commit cf48cce

File: notebooks/query-dependent link analysis.ipynb (hunk 2 above)
Comment (st-alice) [reply notebooks/query-dependent link analysis.ipynb]: Thank you, refactored in commit a3a488951396dda849cfc7c6de72e78d07499605 and 9add08a

File: notebooks/query-dependent link analysis.ipynb (hunk 3 above)
Comment (st-alice) [reply notebooks/query-dependent link analysis.ipynb:333]: Thanks again, refactored in commit a3a488951396dda849cfc7c6de72e78d07499605 and 9add08a

File: notebooks/query-dependent link analysis.ipynb (hunk 4 above)
Comment (st-alice) [reply notebooks/query-dependent link analysis.ipynb]: Thanks again, refactored in commit cbeea18

**Comment by st-alice** [comment]
> Thanks Alice. Just needs some cleaner code in areas, using list comprehensions and dict comprehensions.

Many thanks Jake. I have responded to all the comments.

**Review by JakeRutherford — APPROVED** [review]
All requested changes have been made. Good to go.

**Pull request merged**